test/powershell/Language/Scripting/Requires.Tests.ps1

```diff
@@ -212,3 +311,78 @@ Describe "#requires -Modules" -Tags "CI" {
         }
     }
 }
+
+Describe "#requires warning emitting" -Tags "CI" {
```

Comment:
I am wondering if there is anyway to redirect warning output from the ParseInput function that returns a ScriptBlockAst using 3>&1?
When I try to compare, it is comparing the Extent attribute of ScriptBlockAst rather than the warning message.

Reply:
@vexx32 Would you know how to address the issue I mentioned above? I've taken a deeper look, but I believe that because of the return type ScriptBlockAst of ParseInput, ScriptBlockAst does not have any attribute to get the output warning.

Reply:
Oh, that's a good point. 

Probably the quickest way would just be to invoke the scriptblock. I think you can call `$warnings.GetScriptblock()` to get a fully formed scriptblock and them invoke that directly.

Reply:
@vexx32 Do you mean passing $Warnings into ParseInput? I tried that with redirection, but it still returns a ScriptBlockAst. I also tried using Invoke-Command and passing in the ScriptBlock, but the warnings are not being emitted. It seems to me that we must use the ParseInput function to see the warnings emitted, but the ParseInput function does not seem to work with redirection. Would you know of any way to resolve this?

Reply:
I think you can do something like this:

```powershell

$ast = [System.Management.Automation.Language.Parser]::ParseInput($Statement, [ref]$tokens, [ref]$errors)
$warnings = $ast.GetScriptblock().Invoke() 3>&1
```

If the method necessary isn't available (mightn't be public? Not sure) then you always have the option of:

```powershell
# after checking for parse errors etc
$warnings = & [Scriptblock]::Create($statement) 3>&1
```

Reply:
It seems that the method I am using to print warnings, `PSHostUserInterface.WriteWarningLine` prints out the warning, but does not work with redirection. Is this the standard way to print out warnings in PowerShell, or am I using the wrong function to print warnings?

Reply:
I noticed that when you run a pwsh script from pwsh or when I use the ParseInput function, it prints out the warnings. However, when I try to run it through ScriptBlock.Invoke() or Invoke-Expression command, the warnings are not printed. The warning mechanism is implemented in tokenizer.cs. Would you know if tokenizer.cs gets run when I use ScriptBlock.Invoke() or Invoke-Expression?

Reply:
Yeah, the tokenizer would be called when the script is parsed. If you call [scriptblock]::Create(), that's when the warning would be created. However.. Warnings sent to the usual Warning stream are emitted when code is executed. The code you're trying to implement here appears to be trying to emit warnings when code is _parsed_. So I think you're going to run into an issue where unless the parsing routines are invoked directly from a PowerShell command context, there's nowhere for the warnings to be written.

@daxian-dbw is there a way that we could sensibly write warnings in this instance in a way that we can test that utilises the usual warning stream & is functional in all cases where the tokenizer routines are called?